mrjob/py2.py

```diff
@@ -0,0 +1,170 @@
+
+Strategies for making `mrjob` work across Python versions:
+
+Bytes vs. Unicode
+-----------------
+
+It's tempting to use `from __future__ import unicode_literals` and require
+that all non-byte strings be unicode. But that doesn't really make sense for
+Python 2, where str (bytes) and unicode can be used interchangeably.
+
+So really our string datatypes fall into two categories, bytes, and
+"strings", which are either ``str`` (i.e., bytes) or ``unicode`` in Python 2,
+and ``str`` (i.e. unicode) in Python 3.
+
+These things should always be bytes:
+
+- input data files
+  - use ``'b'`` when opening files: ``open(..., 'rb')``
+  - read data from ``sys.stdin.buffer`` in Python 3, not ``sys.stdin``
+- data encoded by protocols
+- data from subprocesses (this already happens by default)
+- log files parsed by mrjob
+- file content from our filesystem interfaces.
+
+Instead of using ``StringIO`` to deal with these, use ``io.BytesIO``.
+
+Not that both Python 2.6+ and Python 3.3+ have the ``bytes`` type and
+``b''`` constants built-in.
+
+These things should always be strings:
+
+- streams that you print() to (e.g. ``sys.stdout`` if you mock it out)
+- streams that you log to
+- paths on filesystem
+- URIs
+- arguments to commands
+- option names
+- Hadoop counter names and groups
+- Hadoop status messages
+- anything else we parse out of log files
+
+These things are strings because it makes for simpler code:
+
+- contents of config files
+- contents of scripts output by mrjob (e.g. the setup wrapper script)
+- contents of empty files
+
+Use the ``StringIO`` from this module to deal with strings (it's
+``StringIO.StringIO`` in Python 2 and ``io.StringIO`` in Python 3).
+
+Please use ``%`` for format strings and not ``format()``, which is much more
+picky about mixing unicode and bytes.
+
+This module provides a ``basestring`` type so you can check if something
+is a "string".
+
+We don't provide a ``unicode`` constant:
+
+- Use ``not isinstance(..., bytes)`` to check if a string is Unicode
+- To convert ``bytes`` to ``unicode``, use ``.decode('utf_8')`.
+- Python 3.3+ has ``u''`` literals; please use sparingly
+
+If you need to convert bytes of unknown encoding to a "string" (e.g.
+to ``print()`` or log them), you can use ``to_string()``.
+
+Iterables
+---------
+
+Using ``.iteritems()`` or ``.itervalues()`` in Python 2 to iterate over a
+dictionary when you don't need a list is best practice, but it's also (in most
+cases) an over-optimization. We'd prefer clean code; just use ``.items()``
+and ``.values()``.
+
+If you *do* every need that extra efficiency ``for k in some_dict`` does not
+create a list in either version of Python, and you for performance-critical
+code, it's fine to write custom code for each Python version (use ``IN_PY2``
+from this module to check which version you're in).
+
+Same goes for ``xrange``; just use ``range`` (this module provides ``xrange``
+solely to support ``ReprProtocol``).
```

Comment:
If you don't want xrange to be a public interface, should stick it into that one spot it's used.